### PR TITLE
Instrument attributes (accuracy, precision, resolution, make/model, etc)

### PR DESCRIPTION
**Pull request by jessicaaustin**
Rendered version of this branch can be found here:

https://mwengren.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html#variables and:
https://mwengren.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html#instrument

will attempt to keep it up to date with new commits as available.

### CODE REVIEW

**Comment by jessicaaustin** [comment]
Summary of the conversation so far (which has been happening over email):

Immediate motivators are 
a) if NDBC is pulling data from ERDDAP, they need to know salinity accuracy, and 
b) OOI and other projects want to publish instrument information over ERDDAP, and it would be nice to do that in a standard way.

Related standards:

* [NCEI templates](https://www.nodc.noaa.gov/data/formats/netcdf/v2.0/) `instrument`
* [GliderDAC variables](https://ioos.github.io/ioosngdac/ngdac-netcdf-file-format-version-2.html#variables)

The specific fields we're discussing here are

* resolution
* accuracy
* precision
* calibration_date
* make_model

There has been some debate about whether these attributes should be on the *data variable* (e.g., conductivity) or the *instrument variable* (e.g., CTD). As of this writing, we're putting make_model and calibration_date on the instrument, and accuracy, precision, and resolution (RAP) on the data variable. 

The Glider DAC profile puts RAP as attributes on the data variable. NCEI templates puts then on the instrument variable, but I've yet to find a meaningful example where it's actually filled in. And RAP is more specific to the sensor, not the instrument, so that's the main argument for having it on the data variable.

The other thing we talked about was how to keep track of multiple RAP values due to multiple calibrations over time -- we punted on that and said that these attributes should only specify the value since the last calibration date. This approach is simple and doesn't require the addition of new variables, and it's still useful for real-time data (which serves the NDBC/GTS case). It doesn't provide a way to get a history of RAP over time, but we felt that was beyond the scope of what we need to do right now.

**Comment by jessicaaustin** [comment]
@lukecampbell and/or @kwilcox if you have time, could you please review this PR? In particular, do you think these new attributes are in the right place (data var versus instrument var), and do you agree with the definitions? 

You can see rendered versions here:

https://mwengren.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html#variables and:
https://mwengren.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html#instrument

**Comment by lukecampbell** [comment]
The metadata attributes and descriptors are good. I think we might need to communicate to the implementers how the values should be represented though. Resolution, Accuracy, and Precision are often not a linear relation but often a polynomial relation or more complex, and are often dependent on the operating environment or instrument configurations, so I don't know if it's necessarily intuitive or accurate to represent these sensor characteristics as scalar values. A lot of sensor data sheets give ranges of these characteristics that can be expected within standard operating envelopes, or a statistical relation. Maybe we should describe the attributes as "the expected characteristic within the standard operating envelope for **this** sensor." or something to that effect?

**Comment by jessicaaustin** [comment]
@lukecampbell I can think of two different concrete scenarios here:

a) NDBC

Set `accuracy` for salinity variables, to replace the `<fm64k2>` tag. The values you can provide for this tag are:

    0, indicates salinity is not measured
    1, indicates salinity accuracy is better than 0.02 ppt.
    2, indicates salinity accuracy is less than 0.02 ppt.

Note that these are ranges that they want, not absolute values.

b) OOI

Set instrument `make_model` and `calibration_date`, and also include a `references` attribute that points to their DPS sheets (e.g., https://oceanobservatories.org/data-product/pracsal/). Right now there isn't an easy way to parse out the latest RAP info for a sensor so we're not planning on including that in the ERDDAP dataset.

For fun, I looked at the DPS for salinity and here's what it says about accuracy:

>  assuming typical ocean values, 5 deg C and 29 mS/cm conductivity, the resulting practical
salinity inaccuracy is 0.01 at 100m (mainly from conductivity) and 0.005 at 5000m (from
conductivity and pressure) using the algorithm code described in section 4.8

So this is what you're talking about right? Accuracy is a function of depth/pressure, temperature, and even conductivity. But it seems like you could provide a *range* of values?

Basically could we define this as a range of values instead of a single value? Or should we set it as the *worst* value? (e.g., in the OOI salinity example above, the accuracy would be `0.01`). I feel like if the representation is more complex than a single value or a range, we should just refer to external documentation and not try to embed that info into the dataset metadata -- that's what a datasheet is for.

**Comment by lukecampbell** [comment]
It's hard to predict what these metadata could be used for by data consumers, and it's hard to convey the complexities of the sensor characteristics within the metadata attributes. 

For example, if an anemometer has an accuracy of ±0.2 kt when the wind is blowing > 3 kts, but has a highly variable and unpredictable accuracy below 3 kts, how would we communicate that in the metadata?

**Comment by mwengren** [comment]
@jessicaaustin just pinging on status of this.  I know we talked about it at last week's meeting, but forgot what was decided and didn't see anything in meeting notes.  Did we plan to ask for review by anyone else (Luke?), or is it ready to be merged as is?

I don't think I'm clear on how we plan to integrate the variable mapping spreadsheet we've been working on offline into the Profile, but that can be added as a follow on PR to this I think.

**Edit:** After posting this, the most recent comments appeared on my browser tab.  Let me know what your conclusions are and I can adapt the definitions.  I think we need to keep these attributes as plain/generic as possible, but also meet the immediate need for NDBC, keeping in mind also there's no definitions provided in the Glider DAC [netCDF File Format](https://ioos.github.io/ioosngdac/ngdac-netcdf-file-format-version-2.html#variables), which is what we're using as a predecessor.

**Comment by jessicaaustin** [comment]
@mwengren 

> I think we need to keep these attributes as plain/generic as possible

Agreed, and @lukecampbell 's comments only emphasize this.

Re-reading our definition for `accuracy`:

> geophysical_variable:accuracy | IOOS | The sensor accuracy is the closeness of the measurements to the variable's true value. It should be given in the same units as the measured variable. If the instrument has been calibrated multiple times with different results, the most recent accuracy should be provided here

I think this is vague enough? It doesn't say that it should be a single value, or a range, or any specific format. 

For what it's worth, I couldn't find any real-life examples of RAP attributes in the wild, and Luke said he didn't know of any either. So I think adding the fields as-is is a good start, and we can see where it goes -- we can always build upon or narrow the definition later if it makes sense. Right now we have only one concrete example to work with for RAP attributes, and imho this PR solves it.

> I don't think I'm clear on how we plan to integrate the variable mapping spreadsheet we've been working on offline into the Profile, but that can be added as a follow on PR to this I think.

Agreed it's a separate thing. Let's merge in this PR to add the fields, then we should have a separate update that gives instructions on what values to use to send this info to NDBC.

**Comment by mwengren** [comment]
Ok.  Merging.

**Comment by jessicaaustin** [comment]
Here's an example of this in action: http://testing.erddap.axds.co/erddap/tabledap/c10-wfs-central-buoy-25m-isobath.html -- look at the salinity variable.

(note: the link above is for a testing server; at some point these updates will make it to http://erddap.sensors.axds.co/erddap/tabledap/c10-water-velocity-wfs-central-b.html which should continue to exist)

**Comment by lukecampbell** [comment]
To help understanding the sensor characteristics, specifically the complexities of accuracy, I attached the sensor characteristics sheet for a scale I just bought for measuring coffee. It has very different accuracy characteristics depending on it's environment, and the mass of the object being measured.

image
